content/architecture/storage-engine/prolly-tree.md

```diff
@@ -2,38 +2,234 @@
+
+Also, writes to B-trees are not history independent, the order of the writes internally changes the structure of the tree. Thus, storage cannot be easily shared between two versions of the same tree.
+
+## Building a Prolly Tree
+
+The easiest way to understand Prolly trees is to walk through, step-by-step, how we build one. You start with a map of keys to values. Using the above B-tree example, we have keys between 1 and 21. The values don't really matter. Just imagine them dangling off the keys.
+
+1. **Sort**: Sort the map by its key value, so that it is laid out in order. 
+
+![Building a Prolly-Tree, Step 1](../../.gitbook/assets/tim-prolly-tree-step-1.png)
+
+2. **Determine Chunk Boundaries**: Use a seed, the size of the current chunk, and the key value to calculate a rolling hash. Any time the hash value is below a target value, form a chunk boundary and start a new block. Here's the chunking step on our leaf nodes:
+
+![Building a Prolly-tree, Step 2](../../.gitbook/assets/tim-prolly-tree-step-2.png)
+
+3. **Hash each Chunk**: You now have the leaf nodes of the Prolly Tree. Compute the content address of each block by applying a strong hash function to its contents. You store the contents in a content addressed block store. Here our blocks have been addressed and stored in the block store:
+
+![Building a Prolly-tree, Step 3](../../.gitbook/assets/tim-prolly-tree-step-3.png)
+
+4. **Finished?**: If the length of your list of content addresses is 1, then you are done. This is the content address of your tree.
+
+5. **Build a New Map** Otherwise, form the next layer of your Prolly Tree by creating a map of the highest key value in the chunk and the content address of the chunk. Here's what the entries for the first internal level of the tree would look like for our example:
+
+![Building a Prolly-tree, Step 5](../../.gitbook/assets/tim-prolly-tree-step-5.png)
+
+6. **Return to Step 2**: Use this new map step 2. The algorithm terminates when step 4's condition is reached.
+
+Following these steps results in the following Prolly Tree. 
+
+![Prolly Tree](../../.gitbook/assets/tim-prolly-tree-example.png)
+
+## Modifying a Prolly Tree
+
+The magic of Prolly Trees is not seen on construction, but on modification. As you'll see, modifying a Prolly Tree changes the minimal amount of content addresses. Minimal modification means more structural sharing of contents because fewer content addresses change. It is also the basis of fast diff which is accomplished by comparing content addresses.
+
+### Update a Value
+
+If we update a value, we walk the tree by key to the leaf node holding the value. We then edit the referenced chunk in place. After the edit, we recalculate the content address of the chunk. We then walk up the tree recalculating each internal content address up to the root of the tree. 
+
+![Prolly Tree Value Update](../../.gitbook/assets/prolly-tree-update-value.png)
+
```

Comment:
Not sure if it's overly specific to mention this is only true for fixed-size value updates. If you change the length of a value, then it does have to re-chunk.

Instead of saying "in-place", it might be more clear if we said we could create the new leaf chunk by directly modifying the existing value in a copy of the existing chunk. That way it's clear that we're copy-on-write and the old chunk needs to stay around for versioning.